File: c++/UnThreadSafe.cpp
#include <boost/bind.hpp>
#include <boost/thread.hpp>
#include <iostream>

void increment(int *x) {
  for (int i = 0; i < 1000; i++) {
    (*x)++;
    boost::this_thread::sleep(boost::posix_time::milliseconds(1));
  }
}

int main(void) {
  int countNum = 0;

  boost::thread th1(boost::bind(increment, &countNum));
  boost::thread th2(boost::bind(increment, &countNum));
  boost::thread th3(boost::bind(increment, &countNum));
  boost::thread th4(boost::bind(increment, &countNum));

  th1.join();
  th2.join();
  th3.join();
  th4.join();

  std::cout << countNum << std::endl;
}

File: java/src/main/java/com/example/multi_thread/UnThreadSafe.java
package com.example.multi_thread;

public class UnThreadSafe {

    public int countNum = 0;

    public static void main(String[] args) {

        UnThreadSafe main = new UnThreadSafe();

        Runnable runner = new Runnable() {

            public void run() {
                for (int i = 0; i < 1000; i++) {
                    main.countNum++;
                    try{
                        Thread.sleep(2);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        };

        Thread th1 = new Thread(runner);
        Thread th2 = new Thread(runner);
        Thread th3 = new Thread(runner);

        th1.start();
        th2.start();
        th3.start();

        try {
            th1.join();
            th2.join();
            th3.join();
        } catch (Exception e) {
            e.printStackTrace();
        }

        System.out.println(main.countNum);

    }
}